### PR TITLE
[fa] add Supply to fungible_asset

### PR DESCRIPTION
**Pull request by lightmark**
### <samp>🤖 Generated by Copilot at a491605</samp>

Refactored the supply tracking logic of fungible assets in the `aptos-framework` module to use a simpler and more robust design. Replaced nested options in the `Metadata` resource with global `Supply` resources and added error handling for underflow cases. Updated the documentation, code, and tests in `fungible_asset.md` and `fungible_asset.move`.

### CODE REVIEW

**Comment by lightmark** [comment]
@davidiw For FA, it's unnecessary to have two structs if we're sure there will be an upgrade soon.

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 1]
```diff
@@ -167,12 +166,18 @@ module aptos_framework::fungible_asset {
...
                 decimals,
             }
         );
+        if (option::is_some(&monitoring_supply_with_maximum)) {
+            let maximum = option::destroy_some(monitoring_supply_with_maximum);
+            move_to(metadata_object_signer, Supply {
+                current: 0,
+                maximum
+            });
+        };
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/fungible_asset.move:180]: do we offer optional supply in coin? if so, fine, if not, not so great....

**Review by davidiw — COMMENTED** [review]
could we share this out to the team?

overall fine with it, just want that last bit.

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move [hunk 2]
```diff
@@ -59,16 +60,16 @@ module aptos_framework::fungible_asset {
     /// Maximum possible coin supply.
     const MAX_U128: u128 = 340282366920938463463374607431768211455;
 
-    struct Supply has store {
-        current: OptionalAggregator,
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct Supply has key {
+        current: u128,
+        // option::none() means unlimited supply.
         maximum: Option<u128>,
```
[the remark this answers is not in the document]
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:67]: what's the benefit of skipping option?

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 1 above)
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:180]: yes, this is why I follow the pattern.
FYI: https://github.com/aptos-labs/aptos-core/blob/main/aptos-move/framework/aptos-framework/sources/coin.move#L462

Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:180]: We can actually considering always having a supply if we can figure out how to help existing fungible assets migrate to aggregator later on when aggregator is read

**Review by movekevin — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:72]: save a small amount of perf :P

**Review by davidiw — APPROVED** [review]

File: aptos-move/framework/aptos-framework/sources/fungible_asset.move (hunk 2 above)
[the remark this answers is not in the document]
Comment (lightmark) [reply aptos-move/framework/aptos-framework/sources/fungible_asset.move:72]: The pattern I stick to is to use 0 or max for entry functions but keep `option<T>` for unambiguity.

**Pull request merged**